src/parser/column_definition.cpp

```diff
@@ -1,39 +1,174 @@
+		throw InternalException("Type not implemented for ColumnExpressionType");
+	}
+	}
 }
 
 ColumnDefinition ColumnDefinition::Copy() const {
 	ColumnDefinition copy(name, type);
 	copy.oid = oid;
+	copy.storage_oid = storage_oid;
 	copy.default_value = default_value ? default_value->Copy() : nullptr;
+	copy.generated_expression = generated_expression ? generated_expression->Copy() : nullptr;
 	copy.compression_type = compression_type;
+	copy.category = category;
 	return copy;
 }
 
 void ColumnDefinition::Serialize(Serializer &serializer) const {
 	FieldWriter writer(serializer);
 	writer.WriteString(name);
 	writer.WriteSerializable(type);
-	writer.WriteOptional(default_value);
+	writer.WriteField<TableColumnType>(category);
+	if (Generated()) {
+		writer.WriteOptional(generated_expression);
+	} else {
+		writer.WriteOptional(default_value);
+	}
 	writer.Finalize();
 }
 
 ColumnDefinition ColumnDefinition::Deserialize(Deserializer &source) {
 	FieldReader reader(source);
 	auto column_name = reader.ReadRequired<string>();
 	auto column_type = reader.ReadRequiredSerializable<LogicalType, LogicalType>();
-	auto default_value = reader.ReadOptional<ParsedExpression>(nullptr);
+	auto category = reader.ReadRequired<TableColumnType>();
+	auto expression = reader.ReadOptional<ParsedExpression>(nullptr);
 	reader.Finalize();
 
-	return ColumnDefinition(column_name, column_type, move(default_value));
+	switch (category) {
+	case TableColumnType::STANDARD:
+		return ColumnDefinition(column_name, column_type, ColumnExpression(move(expression)));
+	case TableColumnType::GENERATED:
+		return ColumnDefinition(column_name, column_type,
+		                        ColumnExpression(move(expression), ColumnExpressionType::GENERATED));
+	default:
+		throw NotImplementedException("Type not implemented for TableColumnType");
+	}
+}
+
+bool ColumnDefinition::Generated() const {
+	return category == TableColumnType::GENERATED;
+}
+
+//===--------------------------------------------------------------------===//
+// Generated Columns (VIRTUAL)
+//===--------------------------------------------------------------------===//
+static bool ColumnsContainsColumnRef(const vector<ColumnDefinition> &columns, const string &columnref) {
+	for (auto &col : columns) {
+		if (col.name == columnref) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static void StripTableName(ColumnRefExpression &expr, const string &table_name) {
+	D_ASSERT(expr.IsQualified());
+	auto &name = expr.GetTableName();
+	auto &col_name = expr.GetColumnName();
+	if (name != table_name) {
+		throw BinderException("Column \"%s\" tries to reference a table outside of %s", expr.GetColumnName(),
+		                      table_name);
+	}
+	// Replace the column_names vector with only the column name
+	expr.column_names = vector<string> {col_name};
+}
+
+static void VerifyColumnRefs(const string &name, const vector<ColumnDefinition> &columns, ParsedExpression &expr) {
+	if (expr.type == ExpressionType::COLUMN_REF) {
```

Comment:
Here you are (sort of) reimplementing parts of the Binder, can you not use that directly, e.g. by using the CheckBinder?

Reply:
I've removed those parts, now it only checks (and throws) if any of the column references are qualified, that way I don't need to verify and strip table names - which would pretty much be binding